File: sdk/python/src/aegis_sdk/instrument.py
"""Instrumentation — monkey-patches an OpenAI-compatible client so every LLM call
emits a `TraceEvent` to a sink, with zero changes to the caller's call sites.

This is the package version of the Week 0 Block 3 prototype (`week0/instrument_demo.py`),
with the six follow-ups that block's notes flagged for v0.1 now folded in:

  1. `auto_usage`     — streaming calls get `stream_options={"include_usage": True}`
                        injected so the final chunk carries token counts.
  2. `async_client`   — explicit override for async clients that aren't an
                        `openai.AsyncOpenAI` subclass (see §2.3 of NOTES_block3.md —
                        `inspect.iscoroutinefunction` lies, so we check the client type).
  3. full messages    — the whole `messages` list is persisted (via `TraceEvent.from_call`),
                        not just the last turn.
  4. `deinstrument`   — the original `create` is stashed on the wrapper so the patch
                        can be cleanly removed (tests, eval baselines).
  5. tool_calls       — when the model returns a tool call instead of text, the call is
                        serialized into `completion` rather than recorded as blank.
  6. model normalize  — the model id is normalized to the backend cost-catalog
                        convention (`groq/` prefix) so server-side cost lookup hits.

The sink is injected, not hard-coded: the `Aegis` facade (SDK-C) passes its ring
buffer's `put`. That keeps this module free of any transport/threading concerns — it
only captures and hands off.
"""

from __future__ import annotations

import json
import logging
import time
from collections.abc import Callable, Iterator
from typing import Any

from aegis_sdk.trace import TraceEvent

log = logging.getLogger("aegis_sdk")

Sink = Callable[[TraceEvent], None]

# Tags placed on the wrapped `create` so a second instrument() call is a no-op and
# deinstrument() can recover the original. Stored on the function object itself.
_SENTINEL = "_aegis_instrumented"
_ORIGINAL = "_aegis_original"

# Providers whose models the cost catalog keys with a `provider/` prefix. OpenAI and
# Anthropic models are keyed bare, so only this set triggers prefixing.
_PREFIXED_PROVIDERS = {"groq"}


# ---------------------------------------------------------------------------
# Client-shape + identity helpers
# ---------------------------------------------------------------------------

def _is_async_client(client: Any, override: bool | None) -> bool:
    """Decide whether `client` needs the async wrapper.

    An explicit `override` always wins (the escape hatch for custom async clients).
    Otherwise we check `isinstance(client, AsyncAPIClient)` — NOT
    `inspect.iscoroutinefunction`, which the openai SDK's method wrapping defeats
    (NOTES_block3.md §2.3). If openai isn't importable, default to sync.
    """
    if override is not None:
        return override
    try:
        from openai._base_client import AsyncAPIClient  # not in openai's public API
    except Exception:  # noqa: BLE001 — openai optional / internal path moved
        return False
    return isinstance(client, AsyncAPIClient)


def _base_url_of(client: Any) -> str | None:
    """Best-effort read of the client's base URL — used to infer the provider."""
    try:
        return str(client.base_url)
    except Exception:  # noqa: BLE001
        return None


def _resolve_provider(provider: str | None, base_url: str | None) -> str | None:
    """Resolve the provider: explicit arg wins, else infer from the base URL host."""
    if provider:
        return provider
    if not base_url:
        return None
    if "groq.com" in base_url:
        return "groq"
    if "anthropic.com" in base_url:
        return "anthropic"
    if "openai.com" in base_url:
        return "openai"
    return None


def _normalize_model(model: str, provider: str | None) -> str:
    """Map a raw model id onto the backend cost-catalog key convention.

    Already-prefixed ids (anything containing `/`, e.g. a LiteLLM-style
    `groq/llama-3.3-70b-versatile`) pass through untouched. A bare id from a
    prefix-using provider gets the prefix added; everything else stays bare.
    """
    if not model or "/" in model:
        return model
    if provider in _PREFIXED_PROVIDERS:
        return f"{provider}/{model}"
    return model


# ---------------------------------------------------------------------------
# Response extraction
# ---------------------------------------------------------------------------

def _tool_call_to_dict(tc: Any) -> dict[str, Any]:
    """Flatten one tool-call object into a plain dict for JSON serialization."""
    fn = getattr(tc, "function", None)
    return {
        "id": getattr(tc, "id", None),
        "type": getattr(tc, "type", "function"),
        "name": getattr(fn, "name", None),
        "arguments": getattr(fn, "arguments", None),
    }


def _extract_completion(result: Any) -> str:
    """Pull the assistant's output text from a non-streaming response.

    When the model returns a tool call, `message.content` is None — Block 3's prototype
    recorded that as a blank completion. Here we serialize the tool calls instead, so a
    tool-using turn shows *something* in the trace detail view rather than an empty box.
    """
    try:
        message = result.choices[0].message
    except (AttributeError, IndexError, TypeError):
        return ""
    if getattr(message, "content", None):
        return message.content
    tool_calls = getattr(message, "tool_calls", None)
    if tool_calls:
        return json.dumps([_tool_call_to_dict(tc) for tc in tool_calls], default=str)
    return ""


# ---------------------------------------------------------------------------
# Streaming pass-through wrappers
#
# Both are generators that yield every chunk straight through to the caller while
# accumulating trace state on the side — the caller iterates them exactly as it would
# the raw SDK stream. The `finally` fires `on_done` even on early break or mid-stream
# error (partial read), so a trace is always emitted. See NOTES_block3.md §2.1.
# ---------------------------------------------------------------------------

OnDone = Callable[[str, "int | None", "int | None", "str | None"], None]


def _wrap_stream_sync(stream: Any, on_done: OnDone) -> Iterator[Any]:
    parts: list[str] = []
    tokens_in: int | None = None
    tokens_out: int | None = None
    error: str | None = None
    try:
        for chunk in stream:
            if chunk.choices and chunk.choices[0].delta.content:
                parts.append(chunk.choices[0].delta.content)
            if getattr(chunk, "usage", None) is not None:
                tokens_in = chunk.usage.prompt_tokens
                tokens_out = chunk.usage.completion_tokens
            yield chunk
    except Exception as exc:  # noqa: BLE001 — recorded then re-raised below
        error = repr(exc)
        raise
    finally:
        on_done("".join(parts), tokens_in, tokens_out, error)


async def _wrap_stream_async(stream: Any, on_done: OnDone) -> Any:
    parts: list[str] = []
    tokens_in: int | None = None
    tokens_out: int | None = None
    error: str | None = None
    try:
        async for chunk in stream:
            if chunk.choices and chunk.choices[0].delta.content:
                parts.append(chunk.choices[0].delta.content)
            if getattr(chunk, "usage", None) is not None:
                tokens_in = chunk.usage.prompt_tokens
                tokens_out = chunk.usage.completion_tokens
            yield chunk
    except Exception as exc:  # noqa: BLE001 — recorded then re-raised below
        error = repr(exc)
        raise
    finally:
        on_done("".join(parts), tokens_in, tokens_out, error)


# ---------------------------------------------------------------------------
# instrument / deinstrument
# ---------------------------------------------------------------------------

def instrument(
    client: Any,
    *,
    sink: Sink,
    provider: str | None = None,
    auto_usage: bool = True,
    async_client: bool | None = None,
) -> Any:
    """Patch `client.chat.completions.create` to emit a `TraceEvent` per call.

    Patches the *instance*, not the class — a second bare `OpenAI()` in the same process
    is unaffected (eval baselines must not start emitting traces). Calling `instrument`
    twice on the same client is a no-op. Returns `client` so the call can be chained.

    A per-call `aegis_metadata={...}` kwarg is recognized: it is popped before the
    request reaches the provider and attached to that trace's metadata.
    """
    completions = client.chat.completions
    original = completions.create
    if getattr(original, _SENTINEL, False):
        return client  # already instrumented — idempotent

    is_async = _is_async_client(client, async_client)
    resolved_provider = _resolve_provider(provider, _base_url_of(client))

    def _emit(
        *,
        messages: list[dict[str, Any]],
        model: str,
        completion: str,
        tokens_in: int | None,
        tokens_out: int | None,
        started: float,
        status: str,
        streamed: bool,
        error: str | None,
        user_metadata: dict[str, Any] | None,
    ) -> None:
        # The sink (ring buffer) must never break the caller's call — swallow anything.
        try:
            sink(
                TraceEvent.from_call(
                    model=model,
                    messages=messages,
                    completion=completion,
                    tokens_in=tokens_in,
                    tokens_out=tokens_out,
                    latency_ms=(time.perf_counter() - started) * 1000.0,
                    status=status,
                    streamed=streamed,
                    provider=resolved_provider,
                    error=error,
                    user_metadata=user_metadata,
                )
            )
        except Exception:  # noqa: BLE001
            log.exception("aegis: failed to emit trace (call itself is unaffected)")

    def _prepare(kwargs: dict[str, Any]) -> tuple[
        list[dict[str, Any]], str, bool, dict[str, Any] | None
    ]:
        """Read call args, normalize the model, inject stream_options, pop aegis_metadata.

        Mutates `kwargs` in place: removes our private `aegis_metadata` (so it never
        reaches the provider) and, when `auto_usage` is on, ensures streaming calls ask
        for the usage chunk.
        """
        user_metadata = kwargs.pop("aegis_metadata", None)
        messages = list(kwargs.get("messages") or [])
        model = _normalize_model(kwargs.get("model", ""), resolved_provider)
        streamed = bool(kwargs.get("stream"))
        if streamed and auto_usage:
            opts = dict(kwargs.get("stream_options") or {})
            opts.setdefault("include_usage", True)
            kwargs["stream_options"] = opts
        return messages, model, streamed, user_metadata

    if is_async:

        async def wrapped(*args: Any, **kwargs: Any) -> Any:
            messages, model, streamed, user_metadata = _prepare(kwargs)
            started = time.perf_counter()
            try:
                result = await original(*args, **kwargs)
            except Exception as exc:
                _emit(
                    messages=messages, model=model, completion="",
                    tokens_in=None, tokens_out=None, started=started,
                    status="error", streamed=streamed, error=repr(exc),
                    user_metadata=user_metadata,
                )
                raise
            if streamed:
                def on_done(text: str, tin: int | None, tout: int | None, err: str | None) -> None:
                    _emit(
                        messages=messages, model=model, completion=text,
                        tokens_in=tin, tokens_out=tout, started=started,
                        status="error" if err else "ok", streamed=True, error=err,
                        user_metadata=user_metadata,
                    )
                return _wrap_stream_async(result, on_done)
            usage = getattr(result, "usage", None)
            _emit(
                messages=messages, model=model, completion=_extract_completion(result),
                tokens_in=getattr(usage, "prompt_tokens", None),
                tokens_out=getattr(usage, "completion_tokens", None),
                started=started, status="ok", streamed=False, error=None,
                user_metadata=user_metadata,
            )
            return result

    else:

        def wrapped(*args: Any, **kwargs: Any) -> Any:
            messages, model, streamed, user_metadata = _prepare(kwargs)
            started = time.perf_counter()
            try:
                result = original(*args, **kwargs)
            except Exception as exc:
                _emit(
                    messages=messages, model=model, completion="",
                    tokens_in=None, tokens_out=None, started=started,
                    status="error", streamed=streamed, error=repr(exc),
                    user_metadata=user_metadata,
                )
                raise
            if streamed:
                def on_done(text: str, tin: int | None, tout: int | None, err: str | None) -> None:
                    _emit(
                        messages=messages, model=model, completion=text,
                        tokens_in=tin, tokens_out=tout, started=started,
                        status="error" if err else "ok", streamed=True, error=err,
                        user_metadata=user_metadata,
                    )
                return _wrap_stream_sync(result, on_done)
            usage = getattr(result, "usage", None)
            _emit(
                messages=messages, model=model, completion=_extract_completion(result),
                tokens_in=getattr(usage, "prompt_tokens", None),
                tokens_out=getattr(usage, "completion_tokens", None),
                started=started, status="ok", streamed=False, error=None,
                user_metadata=user_metadata,
            )
            return result

    setattr(wrapped, _SENTINEL, True)
    setattr(wrapped, _ORIGINAL, original)
    completions.create = wrapped
    return client


def deinstrument(client: Any) -> bool:
    """Remove the patch from `client`, restoring the original `create`.

    Returns True if a patch was removed, False if the client wasn't instrumented.
    Used by tests and by callers that want a clean client back (e.g. eval baselines).
    """
    completions = client.chat.completions
    current = completions.create
    original = getattr(current, _ORIGINAL, None)
    if original is None:
        return False
    completions.create = original
    return True

File: sdk/typescript/src/instrument.ts
// Instrumentation — monkey-patches an OpenAI-compatible client so every call
// emits a TraceEvent to a sink, with zero changes to the caller's call sites.
//
// TypeScript port of the Python SDK's instrument.py. The reliability flags
// (auto_usage, tool_calls handling, model normalization, deinstrument) match
// the Python equivalent; the sync-vs-async wrapper split is gone because in
// JS/TS `create` is always async — one wrapper covers everything.

import type { ChatMessage, TraceMetadata } from "./trace.js";
import { TraceEvent } from "./trace.js";

/** What the customer passes us — anything that quacks like an OpenAI client.
 *  Structurally typed so users don't need a runtime instanceof check, and so
 *  fakes work in tests without importing the real `openai` package. */
export interface OpenAILike {
  baseURL?: string | URL;
  chat: {
    completions: {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      create: (params: any) => Promise<any>;
    };
  };
}

export type Sink = (event: TraceEvent) => void;

export interface InstrumentOpts {
  sink: Sink;
  /** Override provider inference from baseURL. Drives model-id normalization
   *  for correct server-side cost lookup. */
  provider?: string;
  /** Inject `stream_options.include_usage = true` on streaming calls so the
   *  trace gets token counts. Default true. */
  autoUsage?: boolean;
}

// Symbols on the wrapped function so a second instrument() call is a no-op,
// and deinstrument() can recover the pre-patch method.
const SENTINEL = Symbol.for("aegis.instrumented");
const ORIGINAL = Symbol.for("aegis.original");

interface InstrumentedFn {
  [SENTINEL]?: true;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  [ORIGINAL]?: (params: any) => Promise<any>;
}

// Providers whose models the backend cost catalog keys with a `provider/`
// prefix. OpenAI and Anthropic models are keyed bare; only Groq triggers
// prefixing today (matches cost.py).
const PREFIXED_PROVIDERS = new Set(["groq"]);


// ---------------------------------------------------------------------------
// Provider + model helpers — pure, unit-tested
// ---------------------------------------------------------------------------

function baseUrlOf(client: OpenAILike): string | null {
  try {
    const value = client.baseURL;
    return value ? String(value) : null;
  } catch {
    return null;
  }
}

export function resolveProvider(
  provider: string | undefined,
  baseUrl: string | null,
): string | null {
  if (provider) return provider;
  if (!baseUrl) return null;
  if (baseUrl.includes("groq.com")) return "groq";
  if (baseUrl.includes("anthropic.com")) return "anthropic";
  if (baseUrl.includes("openai.com")) return "openai";
  return null;
}

export function normalizeModel(model: string, provider: string | null): string {
  if (!model || model.includes("/")) return model;
  if (provider && PREFIXED_PROVIDERS.has(provider)) return `${provider}/${model}`;
  return model;
}


// ---------------------------------------------------------------------------
// Response extraction — pulls the right "completion" string out of an OpenAI
// chat completion. Handles the tool-call case: when `message.content` is null
// (the model called a tool instead of writing text), serialize the tool calls
// into the completion field so the trace has something to display.
// ---------------------------------------------------------------------------

interface ToolCallLike {
  id?: string;
  type?: string;
  function?: { name?: string; arguments?: string };
}

interface ChoiceLike {
  message?: { content?: string | null; tool_calls?: ToolCallLike[] };
}

interface ResponseLike {
  choices?: ChoiceLike[];
  // `usage` is sometimes `null` (partial streaming responses without
  // include_usage), sometimes a partial object, sometimes absent entirely.
  usage?: { prompt_tokens?: number; completion_tokens?: number } | null;
}

export function extractCompletion(response: ResponseLike): string {
  try {
    const message = response.choices?.[0]?.message;
    if (!message) return "";
    if (message.content) return message.content;
    const toolCalls = message.tool_calls;
    if (Array.isArray(toolCalls) && toolCalls.length > 0) {
      return JSON.stringify(
        toolCalls.map((tc) => ({
          id: tc.id ?? null,
          type: tc.type ?? "function",
          name: tc.function?.name ?? null,
          arguments: tc.function?.arguments ?? null,
        })),
      );
    }
  } catch {
    /* fall through */
  }
  return "";
}


// ---------------------------------------------------------------------------
// Stream pass-through wrapper. Async generator that yields every chunk on
// through to the caller while accumulating trace state on the side. The
// `finally` fires `onDone` even on early break or mid-stream error.
// ---------------------------------------------------------------------------

interface StreamChunkLike {
  choices?: Array<{ delta?: { content?: string | null } }>;
  usage?: { prompt_tokens?: number; completion_tokens?: number };
}

type OnDone = (
  text: string,
  tokensIn: number | null,
  tokensOut: number | null,
  error: string | null,
) => void;

async function* wrapStream<T extends StreamChunkLike>(
  stream: AsyncIterable<T>,
  onDone: OnDone,
): AsyncGenerator<T, void, undefined> {
  const parts: string[] = [];
  let tokensIn: number | null = null;
  let tokensOut: number | null = null;
  let error: string | null = null;
  try {
    for await (const chunk of stream) {
      const delta = chunk.choices?.[0]?.delta;
      if (delta?.content) {
        parts.push(delta.content);
      }
      const usage = chunk.usage;
      if (usage) {
        tokensIn = usage.prompt_tokens ?? null;
        tokensOut = usage.completion_tokens ?? null;
      }
      yield chunk;
    }
  } catch (e) {
    error = String(e);
    throw e;
  } finally {
    onDone(parts.join(""), tokensIn, tokensOut, error);
  }
}


// ---------------------------------------------------------------------------
// instrument / deinstrument
// ---------------------------------------------------------------------------

export function instrument<T extends OpenAILike>(client: T, opts: InstrumentOpts): T {
  const completions = client.chat.completions;
  const original = completions.create as InstrumentedFn & typeof completions.create;
  if (original[SENTINEL]) {
    return client; // already wrapped — idempotent
  }

  const sink = opts.sink;
  const autoUsage = opts.autoUsage ?? true;
  const resolvedProvider = resolveProvider(opts.provider, baseUrlOf(client));

  /** Send to the sink; never let a broken sink break the caller's call. */
  const safeSink = (event: TraceEvent): void => {
    try {
      sink(event);
    } catch {
      // The instrumentation must never propagate sink errors back to the
      // user's app. Logging happens at the facade layer (TS-SDK-C).
    }
  };

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const wrapped = async function (this: unknown, params: any): Promise<any> {
    // Pull our private metadata kwarg before the request leaves us. If a user
    // sets `aegis_metadata`, we want it on the trace but not on the wire to
    // the provider — OpenAI would 400 on the unknown field.
    const aegisMetadata: TraceMetadata | undefined = params?.aegis_metadata;
    const cleanParams = { ...(params ?? {}) };
    delete cleanParams.aegis_metadata;

    const messages: ChatMessage[] = Array.isArray(cleanParams.messages)
      ? cleanParams.messages
      : [];
    const rawModel: string = cleanParams.model ?? "";
    const normalizedModel = normalizeModel(rawModel, resolvedProvider);
    const streamed = Boolean(cleanParams.stream);

    if (streamed && autoUsage) {
      // setdefault-style merge: never clobber a user-supplied value.
      const existing = cleanParams.stream_options ?? {};
      cleanParams.stream_options = {
        ...existing,
        include_usage: existing.include_usage ?? true,
      };
    }

    const t0 = performance.now();
    let result;
    try {
      result = await original.call(this, cleanParams);
    } catch (e) {
      safeSink(
        new TraceEvent({
          model: normalizedModel,
          messages,
          completion: "",
          tokensIn: null,
          tokensOut: null,
          latencyMs: performance.now() - t0,
          status: "error",
          streamed,
          provider: resolvedProvider,
          error: String(e),
          userMetadata: aegisMetadata,
        }),
      );
      throw e;
    }

    if (streamed) {
      // Wrap the stream as a yields-through async generator. Caller iterates
      // it exactly as they would the raw OpenAI Stream — chunks pass on
      // unchanged; the trace fires when the iterator exhausts (or breaks).
      return wrapStream(result, (text, tokensIn, tokensOut, error) => {
        safeSink(
          new TraceEvent({
            model: normalizedModel,
            messages,
            completion: text,
            tokensIn,
            tokensOut,
            latencyMs: performance.now() - t0,
            status: error === null ? "ok" : "error",
            streamed: true,
            provider: resolvedProvider,
            error,
            userMetadata: aegisMetadata,
          }),
        );
      });
    }

    // Non-streaming.
    const usage = (result as ResponseLike).usage;
    safeSink(
      new TraceEvent({
        model: normalizedModel,
        messages,
        completion: extractCompletion(result as ResponseLike),
        tokensIn: usage?.prompt_tokens ?? null,
        tokensOut: usage?.completion_tokens ?? null,
        latencyMs: performance.now() - t0,
        status: "ok",
        streamed: false,
        provider: resolvedProvider,
        userMetadata: aegisMetadata,
      }),
    );
    return result;
  } as InstrumentedFn & typeof completions.create;

  wrapped[SENTINEL] = true;
  wrapped[ORIGINAL] = original;
  completions.create = wrapped;
  return client;
}

export function deinstrument(client: OpenAILike): boolean {
  const completions = client.chat?.completions;
  const current = completions?.create as InstrumentedFn | undefined;
  const original = current?.[ORIGINAL];
  if (!original) return false;
  completions.create = original;
  return true;
}
